### ISSUE TITLE
[BBC][new target]Compile with no C lib?

### ISSUE THREAD

**Issue by Fabrizio-Caruso** [cc65/cc65#579]
Hi!

Is it possible to compile some code without specifying the target? Can I compile a code that does not depend on any lib?
There is no lib for the BBC. Is its cfg correct? Can I compile a program for the BBC that does not depend on any lib? And use the cfg found in CC65?

**Comment by oliverschmidt** [comment]
I have no idea about the state of the BBC config file. The BBC is not a supported cc65 target. You correctly noticed that there's no C library for the BCC - that's because it isn't supported.

You can compile (and even link ;-) code without specifying the target by using the `none` target. Check out the `none.cfg`  to see what variables you need to set on the linker command line to control the memory ranges used.

**Comment by prandeamus** [comment]
There is no native library for BBC (athough I think there are some #defines in the tool source for BBC - or maybe I imagined that?).  I was thinking a lot about this recently. The good thing is that there is a well-defined set of entry points in the BBC MOS (OSBYTE and all that) and the code would be common across all the BBC platforms including Master, Electron, and the various tube 6502 processors. On the other hand, the memory configurations vary a lot - e.g. the low address for RAM where most user programs would reside, which corresponds to the LOMEM and PAGE settings in BBC basic. The upper memory limit for conventional RAM is &7FFF for non-tube BBCs but different when tube is active and so on. My sense is that the code for these variants would all be pretty much the same, but the linker configuration files would have many variants. Which is OK, but it's a question of deciding how to manage them.

Secondary point, the BBC casette and disk filing systems have load address and run address metadata that are stored in the disk directory and not part of the file itself. Assuming you want to copy cc65 output files onto a pseudo disk drive used by a BBC emulator, or even onto a real BBC 5" disk, you need to think about that. I'm sure there are ways around this like stub BBC BASIC boot loaders. 

I think BBC ports would be great if there were a standard relocatable format for the file, and a stub to dynamically load the image at the "best" address. This is just my personal opnion. I used to work on BCPL applications for the beeb in the 80s (well, mixed BCPL and assembler). That used a language-specific relocatable format which worked well.

It's not essential to have paged ROM/RAM support for an initial port, but the later BBC models non-tube configurations did have the ability to load code into 16 RAM slots, each of which located in the 16K from &8000 upwards, the MOS occupying the top 16K.

For zero page use, the essential CC65 runtime ZP can be located at &70 which is the safest approach if you want to generate libraries that can be used and shared with BBC Basic apps. If you don't need this, ZP can be placed somewhere else. The BBC Advanced User Guide has all the memory layouts (And I think is also freely downloadable now if you google it).

Some existing platforms have an EM library for extended memory, but I don't know if this can be made compatible with BBC MOS. 

Tube processors use 65C02 I think, as also does BBC Master? BBC original models A and B are plain old 6502 even if clocked quite fast for the time at 2Mhz.

I think the best approach is to develop some test applications with the none target as mentioned above, and maybe implement a few simple apps that way and get a feel for it. I'd like to see that sort of thing and might be able to collaborate.

The best emulator Ive found has been BeebEm, which was last updared in 2012 but works nicely under windows. There are BBC emulators that work under Linux but I have not found anything as flexible or accurate. (It even emulates the exact buzz and click noises from the floppy drives). Alas, I sold my old BBC for a pittance in the 90s and I should have kept it!

**Comment by oliverschmidt** [comment]
> My sense is that the code for these variants would all be pretty much the same, but the linker configuration files would have many variants. Which is OK, but it's a question of deciding how to manage them.

Linker configuration files can use parameters given on the linker command line. You can put the parameters and their useful values in the docs. See http://cc65.github.io/doc/apple2.html#ss4.1 as an example.

**Comment by groessler** [comment]
Years ago I created an assembler "hello world" for BBC using the BBC linker config file. It worked. I was using a tool called "bbcim" to create an *.INF file, which contains the additional information for the directory entry. If someone is interested I can post the program.

**Comment by Fabrizio-Caruso** [comment]
@groessler, anything that can make CC65 more usable with yet non-supported targets, is welcome.

@all, are the libs the only thing that is missing for a complete support of the BBC system?
Could I use the BBC .cfg file with the no-target option?
If the BBC .cfg file is correct, I could still use CC65 to compile my game if I manage to implement whatever is necessary for input and output. Am I right on this?

**Comment by mrdudz** [comment]
yes, start with a trivial program that just increments some color register or so (so you can see it runs) and then add missing library stuff as needed

**Comment by Fabrizio-Caruso** [comment]
@mrdudz , that was my plan but before doing this I want to make sure I can produce some binary that I can load and run somehow. That is why I have asked about the old unsupported (but still present) .cfg file. If the .cfg is correct, I can be confident that at least the binary is well-formed.
I also need to figure out which emulator has the right option to load it into its memory.

**Comment by oliverschmidt** [comment]
> Could I use the BBC .cfg file with the no-target option?

Likely. However I'm not sure if the file is maintained. However the file doesn't contain too much magic. Basically it only says that the zero page locations to use are $70ff and that the main memory to use is $E00 - $8000.

I personally would rather use the `none.lib` together with the `none.cfg` file and set the values in question via linker command line parameters: `-S $E00 -D __ZPSTART__=$70 -D __STACKSTART__=$8000`.

> If the BBC .cfg file is correct, I could still use CC65 to compile my game if I manage to implement whatever is necessary for input and output. Am I right on this?

That's the very use case of the `none` target.

Regards,
Oliver

**Comment by prandeamus** [comment]
From memory, having base memory starting at E00 is a characteristic of cassette based file systems. $1900 I think is more conservative for a quick test program, depending on your emulator. ZP from 70-8F (not FF) are not used by BASIC therefore least likely to cause problems. $8000 is the first rom byte in a model B 32K system. Ram goes up to $7FFF.

The principle that Oliver suggests seems sound, might need to change a couple of numbers to be good to execute on all common environments.

**Comment by prandeamus** [comment]
A trivial program might be to load the address of a string containing CAT and passing that to OSCLI - this will force the program to execute *CAT and there list the directory of the current floppy disk. I think MOS uses zero terminated strings like C .

**Comment by oliverschmidt** [comment]
Me:
> zero page locations to use are $70ff

@prandeamus:
> 70-8F (not FF)

I wasn't referring to $70-$FF. Rather I was referring to https://en.wiktionary.org/wiki/ff.

Me:
> `-D __STACKSTART__=$8000`

@prandeamus:
> $8000 is the first rom byte in a model B 32K system. Ram goes up to $7FFF.

Then `-D __STACKSTART__=$8000` is correct.

**Comment by oliverschmidt** [comment]
> Call address &FFF7
X and Y should point to a line of text (X=low-byte, Y=high-byte).

That could be archived in pure C without hacking with _sys() -  http://cc65.github.io/doc/funcref.html#ss3.14.

**Comment by prandeamus** [comment]
> Rather I was referring to https://en.wiktionary.org/wiki/ff.

Oh, I see. Thanks,

**Comment by prandeamus** [comment]
> I also need to figure out which emulator has the right option to load it into its memory.

For BeebEm on Windows, I don't think you can do it directly. You have to find a utility that can create and edit BBC disk images, write the executable to that image, and then start BeebEm to auto load that disk image. Your mileage may vary

**Comment by Fabrizio-Caruso** [comment]
@groessler, I am interested in whatever you used to build your hello world example for the BBC yet-to-become-target.

**Comment by prandeamus** [comment]
I would also be interested.

**Comment by groessler** [comment]
I've published it at ftp://ftp.groessler.org/pub/chris/cc65/bbc-hello.tar.gz .
It was first test since I toyed with the idea of adding BBC support to the runtime lib. But I never came around to implement it...

**Comment by Fabrizio-Caruso** [comment]
Thanks @groessler !
I think that just having a working cfg and some pointers to structs (or other getters and setters) to access the hardware for input and output, would be enough for the initial support of a target.
This would encourage people to use it and maybe write some portions of the libs.

**Comment by groessler** [comment]
> I think that just having a working cfg and some pointers to structs (or other getters and setters) to access the hardware for input and output, would be enough for the initial support of a target.

You're free to add a Pull-Request for those... :-)

**Comment by Fabrizio-Caruso** [comment]
@groessler, once I get to work the BBC pseudo-target, I will more than happy to share whatever I find out.

**Comment by 8bit-Dude** [comment]
Hey guyz, sorry to dig up an old thread but were you aware of this:
https://github.com/dominicbeesley/cc65
It seems that dominic did most of the work towards implementing BBC.lib, and that perhaps this branch could be integrated into the CC65 trunk???

**Comment by prandeamus** [comment]
It is certainly interesting, but right now I have no availability. Sorry!
    On Wednesday, 17 July 2019, 04:48:51 BST, Anthony Beaucamp <notifications@github.com> wrote:  
 
 
Hey guyz, sorry to dig up an old thread but were you aware of this:
https://github.com/dominicbeesley/cc65
It seems that dominic did most of the work towards implementing BBC.lib, and that perhaps this branch could be integrated into the CC65 trunk???

**Comment by mrdudz** [comment]
As always its a question of someone willing to spend some time on it... it will certainly need some work :)

**Comment by 8bit-Dude** [comment]
It would sure be great if Dominic could find the time to finish this off... :-)
